### PR TITLE
PINNModel ported to PyTorch

### PR DESCRIPTION
**Pull request by spellsharp**
## Description

Fix #(issue)

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [x] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [x] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/torch_models/tests/test_pinns.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    pinn.fit(dataset, nb_epoch=100)
+
+    x_test = torch.linspace(0, 1, 100).reshape(-1, 1)
+    with torch.no_grad():
+        y_pred = pinn.predict(x_test)
+
+    y_true = x_test  # analytical solution: u(x) = x
+    mse = torch.mean((y_pred - y_true)**2)
+
+    print(f"Final MSE: {mse.item()}")
```
[the remark this answers is not in the document]
Comment (GreatRSingh) [reply deepchem/models/torch_models/tests/test_pinns.py:138]: I noticed there are some more print statements in the test. Could you please remove them when you get a chance? It’ll help keep CI clean and predictable. Thanks a lot!

File: deepchem/models/torch_models/pinns_model.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+                        continue
+
+        return boundary_loss
+
+    def _default_eval_fn(self, x: torch.Tensor) -> torch.Tensor:
+        """Default evaluation function for inference."""
+        with torch.no_grad():
+            return self.model(x)
+
+    def _loss_fn(self, outputs: list, labels: list,
```
Comment (GreatRSingh) [inline deepchem/models/torch_models/pinns_model.py:163]: can you pls add more detail about this loss function, scientific details about why (and how) this is used/helpful etc etc

File: deepchem/models/torch_models/pinns_model.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        self.net = nn.Sequential(nn.Linear(in_channels, 64), nn.Tanh(),
+                                 nn.Linear(64, 64), nn.Tanh(), nn.Linear(64, 1))
+
+    def forward(self, x):
+        if not isinstance(x, torch.Tensor):
+            x = torch.tensor(x, dtype=torch.float32)
+        return self.net(x)
+
+
+class PINNModel(TorchModel):
```
Comment (GreatRSingh) [inline deepchem/models/torch_models/pinns_model.py:25]: Pls add some citations and details about this model.

File: deepchem/models/torch_models/pinns_model.py (hunk 2 above)
Comment (spellsharp) [reply deepchem/models/torch_models/pinns_model.py:163]: Done.

File: deepchem/models/torch_models/tests/test_pinns.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (spellsharp) [reply deepchem/models/torch_models/tests/test_pinns.py:138]: Done

File: deepchem/models/torch_models/pinns_model.py (hunk 3 above)
Comment (spellsharp) [reply deepchem/models/torch_models/pinns_model.py:10]: Done.

File: deepchem/models/torch_models/pinns_model.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+            Dataset to make predictions on
+
+        Returns
+        -------
+        Predicted values
+        """
+        return self.eval_fn(dataset)
+
+
+class NeuralNet(nn.Module):
```
[the remark this answers is not in the document]
Comment (spellsharp) [reply deepchem/models/torch_models/pinns_model.py]: Done!

File: deepchem/models/torch_models/pinns_model.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class PINNModel(TorchModel):
+    """
+    This model is designed for solving linear partial differential equations (PDEs) using
+    Physics-Informed Neural Networks (PINNs). It extends the TorchModel class, and its
+    methods are similar to those in TorchModel, with additional functionality for handling
+    physics-based constraints.
```
Comment (rbharath) [inline deepchem/models/torch_models/pinns_model.py:16]: Add a general usage example

Comment (rbharath) [reply deepchem/models/torch_models/pinns_model.py:16]: An end to end example of how you use PINNModel

File: docs/source/api_reference/models.rst [hunk 6]
```diff
@@ -576,6 +576,12 @@ TextCNNModel
 .. autoclass:: deepchem.models.torch_models.TextCNNModel
   :members:
 
+PINNModel
+---------
+
+.. autoclass:: deepchem.models.torch_models.PINNModel
```
Comment (rbharath) [inline docs/source/api_reference/models.rst:582]: Make sure to update the model cheatsheet as well

**Review by rbharath — COMMENTED** [review]
A few improvements on docs needed but this is pretty close to merge-ready.

**Comment by spellsharp** [comment]
Screenshots of the docstring

File: docs/source/api_reference/models.rst (hunk 6 above)
Comment (spellsharp) [reply docs/source/api_reference/models.rst:582]: Done

File: deepchem/models/torch_models/pinns_model.py (hunk 5 above)
Comment (spellsharp) [reply deepchem/models/torch_models/pinns_model.py:16]: Done.

**Review by rbharath — APPROVED** [review]
LGTM
